src/plugins/hidsim/gui_monitor.cpp

```diff
@@ -0,0 +1,443 @@
+     * publications/2016/05/2x1MicrosoftBug-Economou.pdf, page 11
+     */
+    bool success = register_trap(drakvuf, win32k_json, TRAP_FUNC,
+            &gui_trap, on_draw);
+    json_object_put(win32k_json);
+
+    if (!success)
+    {
+        PRINT_DEBUG("[HIDSIM] [MONITOR] Failed to trap NtUserShowWindow\n");
+        return -1;
+    }
+
+    PRINT_DEBUG("[HIDSIM] [MONITOR] Successfully initialized GUI reconstruction\n");
+
+    return 0;
+}
+/*
+ * Worker function to monitors and reconstruct the GUIs, when NtUserShowWindow-
+ * calls have occured
+ */
+int gui_monitor(drakvuf_t drakvuf, volatile sig_atomic_t* coords,
+    volatile sig_atomic_t* has_to_stop)
+{
+    PRINT_DEBUG("[HIDSIM] [MONITOR] Started GUI reconstruction thread\n");
+
+    int res = -1;
+    vmi_instance_t vmi;
+    struct desktop d = {};
+
+    vmi = drakvuf_lock_and_get_vmi(drakvuf);
+
+    /*
+     * The PID, provided by the desktop struct, is required for accessing the
+     * desktop heap in user mode. Since explorer.exe, dwm.exe or other always
+     * existing processes are used due to their low-order PIDs, those are fairly
+     * stable, so it has not to be recreated in each iteration
+     */
+    if (VMI_FAILURE == find_first_active_desktop(vmi, &d))
+    {
+        drakvuf_release_vmi(drakvuf);
+        return -1;
+    }
+    drakvuf_release_vmi(drakvuf);
+
+    *coords = 0;
+
+    /* Keeps track of time to meet delays */
+    struct timeval curr;
+    struct timeval diff;
+    long elapsed = 0;
+
+    uint32_t x, y;
+
+    struct wnd btn = {};
+
+    while (!*has_to_stop)
+    {
+        if (!has_gui_update)
+        {
+            usleep(DELAY);
+            continue;
+        }
+
+        /*
+         * Checks how many usecs have been elapsed since the last
+         * NtUserShowWindow-call.
+         */
+        gettimeofday(&curr, NULL);
+
+        lock.lock();
+        timersub(&curr, &last, &diff);
+        lock.unlock();
+
+        elapsed = diff.tv_sec * 1000000 + diff.tv_usec;
+
+        /*
+         * Delays the reconstruction for DELAY usecs to wait
+         * until all buttons and windows are drawn
+         */
+        if (elapsed >= DELAY)
+        {
+            PRINT_DEBUG("[HIDSIM] [MONITOR] Detected GUI update\n");
+
+            vmi = drakvuf_lock_and_get_vmi(drakvuf);
+            res = scan_for_clickable_button(vmi, &d, &btn);
+            drakvuf_release_vmi(drakvuf);
+            has_gui_update = false;
+
+            if (res >  0)
+            {
+                /* Sets coordinates to click next */
+                x = btn.r.x0 + btn.r.w/2;
+                y = btn.r.y0 + btn.r.h/2;
+                *coords = x << 16 | y;
+
+                printf("[HIDSIM] [MONITOR] Found \"%S\"-button to click at"
+                    "(%d, %d)\n", btn.text, btn.r.x0, btn.r.y0);
+
+            }
+            /* Some error occured */
+            else if (res < 0)
+            {
+                break;
```

Comment:
`PRINT_DEBUG` here or it will be a nightmare to debug

Reply:
Added a `PRINT_DEBUG` statement.